go/database/mpt/verification_proof.go

```diff
@@ -0,0 +1,359 @@
+func verifyProofArchiveTrie(ctx context.Context, trie *ArchiveTrie, from, to int, observer VerificationObserver) error {
+	blockHeight, empty, err := trie.GetBlockHeight()
+	if err != nil {
+		return err
+	}
+	if empty {
+		return nil
+	}
+
+	if to > int(blockHeight) {
+		to = int(blockHeight)
+		observer.Progress(fmt.Sprintf("setting a maximum block height: %d", blockHeight))
+	}
+
+	observer.Progress(fmt.Sprintf("Verifying total block range [%d;%d]", from, to))
+	for i := from; i <= to; i++ {
+		trie, err := trie.getView(uint64(i))
+		if err != nil {
+			return err
+		}
+
+		observer.Progress(fmt.Sprintf("Verifying blcok: %d ", i))
+		if err := verifyProofLiveTrie(ctx, trie, observer); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// VerifyProofLiveTrie verifies the consistency of witness proofs for a live trie.
+// It reads the trie for the head block and loads accounts and storage slots.
+// It extracts witness proofs for these accounts and its storage,
+// and checks that values in the proof and the database match.
+func VerifyProofLiveTrie(ctx context.Context, dir string, config MptConfig, observer VerificationObserver) error {
+	trie, err := OpenFileLiveTrie(dir, config, NodeCacheConfig{})
+	if err != nil {
+		return err
+	}
+
+	observer.StartVerification()
+	err = errors.Join(
+		verifyProofLiveTrie(ctx, trie, observer),
+		trie.Close(),
+	)
+	observer.EndVerification(err)
+	return err
+}
+
+func verifyProofLiveTrie(ctx context.Context, trie *LiveTrie, observer VerificationObserver) error {
+	rootHash, hints, err := trie.UpdateHashes()
+	if hints != nil {
+		hints.Release()
+	}
+	if err != nil {
+		return err
+	}
+
+	observer.Progress("Collecting and Verifying proof... ")
+	visitor := proofVerifyingVisitor{
+		ctx:      ctx,
+		rootHash: rootHash,
+		trie:     trie,
+		observer: observer,
+	}
+	if err := trie.VisitTrie(&visitor); err != nil || visitor.err != nil {
+		return errors.Join(err, visitor.err)
+	}
+
+	const numAddresses = 1000
+	return verifyProofEmptyAccount(trie, rootHash, numAddresses, observer)
+}
+
+// verifyProofEmptyAccount verifies the consistency of witness proofs for empty accounts that are not present in the trie.
+func verifyProofEmptyAccount(trie *LiveTrie, rootHash common.Hash, numAddresses int, observer VerificationObserver) error {
+	observer.Progress(fmt.Sprintf("Veryfing %d empty addresses...", numAddresses))
+	addresses, err := generateUnknownAddresses(trie, numAddresses)
+	if err != nil {
+		return err
+	}
+	for _, addr := range addresses {
+		proof, err := trie.CreateWitnessProof(addr)
+		if err != nil {
+			return err
+		}
+		// expect an empty account
+		if err := verifyAccountProof(rootHash, proof, addr, AccountInfo{}); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// verifyProofEmptyStorage verifies the consistency of witness proofs for empty storage that are not present in the trie.
+func verifyProofEmptyStorage(trie *LiveTrie, rootHash common.Hash, addr common.Address) error {
```

Comment:
"Empty Storage" is a property of an account; in particular, this is the case if the storage root of an account is the empty node.

I think this function should be renamed to "verifyProofsForUnusedStorageSlots"

Reply:
ok, done